decisions/DR009-Batch-method-result-types.md

```diff
@@ -0,0 +1,272 @@
+accessible via the git history).
+
+When deciding on the performance impact of one design over another, we
+had several concerns
+
+1. Our main concern: is it cheaper to call a callback repeatedly, once
+   for each result, or is it cheaper to return an array of results and
+   iterate over it?
+2. Assuming the user _wants_ an array of results, what is the cost of
+   using a callback design anyway, and have the user construct their own
+   array?
+3. Is there a performance benefit to using raw function pointers rather
+   than `std::function`?
+4. What is the performance impact of the small-size optimization in
+   common `std::function` implementations?
+
+### Test setup
+
+The following results come from a build using with GCC 9.3 on Ubuntu
+20.04, compiled in a Release (`O3`) configuration.
+
+Each set of cases is executed over 5000 iterations (epochs), with a
+warm-up of 1000 epochs to alleviate CPU scaling artifacts, etc. In
+addition, the whole test is executed 5 times (5 runs of the perftest
+executable) and the summary results with the lowest standard deviation
+taken, on the assumption that this represents the results that are least
+skewed by external factors.
+
+A database (simple map of entity reference to URL) of size 10000 is
+generated to draw from. Input entity references are randomly chosen from
+this database, with the probability of an erroneous entity reference set
+to 0.1. Timings are nanosecond precision.
+
+Since we're concerned with the relative performance, the key output of
+the perftest is the ratio of timings between two solutions. The mean and
+standard deviation of these timing ratios are reported in the tables
+below. We round to 3 significant figures - the resolution of the fastest
+results are 3 significant figures (hundreds of nanoseconds).
+
+As shorthand for the different test cases, we define
+* _vector_ as the case that returns an array of results.
+* _callback_ as the case that executes a `std::function` callback on
+  each result.
+* _callbackToVector_ as the case that builds an array as callbacks
+  are executed and then iterates over it.
+* _callbackFnPtr_ as the case that uses function pointers, rather than
+  `std::function`, as the callback.
+* _callbackLargeCapture_ as the case that uses a `std::function`
+  callback holding a lambda with capture data such that the small-size
+  optimisation is unavailable.
+
+### Batch size = 1
+
+The following tables gives the results of the (likely) usual case of a
+batch size of 1 (i.e. only querying a single entity reference).
+
+Firstly, the primary results comparing the array return value solution
+versus the `std::function` callback solution is shown below.
+
+| numerator/denominator     | mean | std dev |
+|---------------------------|------|---------|
+| vector/callback           | 1.29 | 0.761   |
+| vector/callbackToVector   | 1.15 | 0.570   |
+| callbackToVector/callback | 1.24 | 0.745   |
```

Comment:
[question] Do you think this is a useful comparison to have in the table (at the risk of shadowing the conclusion that the callback method is faster, even when building a vector?) Same applies to the larger batch results.

Reply:
Good point, this comparison isn't discussed below here, so should be removed.

Reply:
Removed in 73ccbccd